cpp/src/arrow/compute/kernels/scalar_if_else.cc

```diff
@@ -1413,6 +1417,588 @@ struct CaseWhenFunctor<NullType> {
   }
 };
 
+static Status ExecVarWidthScalarCaseWhen(KernelContext* ctx, const ExecBatch& batch,
+                                         Datum* out) {
```

Comment:
Why not return `Result<datum>`?

Reply:
It's minor but 1) we use the shape of out to determine what to return below and 2) it fits slightly better with the rest of the APIs as they all use the `Status(..., Datum* out)` pattern.